### ISSUE TITLE
data-saferedirecturl defeats URL stripping

### ISSUE THREAD

**Issue by aspiers** [newhouse/url-tracking-stripper#69]
I see that Gmail are doing something really evil with hyperlinks within emails:

    <a href="https://link.to.other.site" target="_blank" data-saferedirecturl="https://www.google.com/url?q=link.to.other.site&amp;source=gmail&amp;ust=1534942837200000&amp;usg=AFQjCNH3Pzb1Mq6zL847zzh6iqd4g1B3IA">here</a>

This is also described here:

- Rob--W/dont-track-me-google#2
- [security - Why do links in gmail redirect? - Stack Overflow](https://stackoverflow.com/questions/5172168/why-do-links-in-gmail-redirect)

I really don't want google tracking which links I click on.  I'm sure Google would claim that they're doing the user a favour by hiding referrer data from the target web server, but only because they're stealing that data for themselves!

So I think it would be great if this project stripped these `data-saferedirecturl` links.

**Comment by aspiers** [comment]
Also worth noting that this creates the exact opposite effect of the problem described in #68: the stripped URL is shown on mouse-over, but the hijacked URL is actually used on click.

**Comment by cameronrcraig** [comment]
This extension already handles these kinds of links.

https://www.google.com/url?q=https://website.com causes a 307 redirect (because of the extension) to https://website.com

The rule is:

{
    name: 'Google Search Results',
    targetParam: 'url',
    patterns: [
      `${SCHEMA}www.google.com/url?`
    ],
    // Google uses 'ping' method sometimes.
    types: ['main_frame', 'ping']
  }

**Comment by cameronrcraig** [comment]
Let me know if my changes fixed it, if not I'll look into it further!

Also I'm not crazy familiar with git but I appreciate the tips, I wasn't sure if I should submit pull requests or continue developing on my own fork because I don't know if the original developer is still around to accept those pull requests.

**Comment by YesThatAllen** [comment]
I wonder what this PR was supposed to do?

https://github.com/newhouse/url-tracking-stripper/pull/23

What I see, when using the official branch, is that an email from github which I receive in gmail's webmail is still wrapped in a `data-saferedirecturl`. When I click the link, I suffer through google's link-tracking (adding an undesired hop, and tracking my actions) then land at the github issue I wanted to visit.

It's only when I'm on the github site that the URL Tracking Stripper extension claims to have helped (It hasn't) and offers to "Reload with full URL".. 

![2018-09-26_09-46-40](https://user-images.githubusercontent.com/2879972/46088026-2588fd00-c171-11e8-829b-ad573f73a657.png)

(Nothing was skipped here)

**Comment by newhouse** [comment]
TLDR: It *does* skip Google and does *not* pass through its servers. Please let me know if you can show otherwise and how to reproduce it, and I will certainly investigate @YesThatAllen. This will be the last time I comment on this thread. If there still turns out to be some issue, please open another ticket as this one has gotten quite useless.

Details:
Let me see if I can put this to bed once and for all: 
- As mentioned several times now, I will not go into your webpages and manipulate links as this is very invasive. Because of this, the `data-saferedirecturl`s all exist in Gmail and wherever you encounter them - I do not change them on the webpages and if you look at the HTML they will still be there. This extension is not intended to change the URLs on the page for you, so this is expected behavior. But it also doesn't mean that it's not going to skip those redirects for you...
- When you click on one of them with this extension installed, the extension will BLOCK that request and redirect you to the target page, skipping Google's servers and having no cookies dropped or tracking occur. You may see the URL flash in your browser bar or network history, but it will have been stopped from loading via a `307 Internal Redirect`.
- The following videos illustrate this behavior by navigating to a link pulled from a `data-saferedirecturl` first with the extension enabled, and second without it enabled:
  - [With extension enabled](https://www.screencast.com/t/Baq183nqx1Y)
  - [Without extension enabled](https://screencast.com/t/fAFTQpItDk)

Here's an image as well. 
![image](https://user-images.githubusercontent.com/2524009/46090640-00db5800-c166-11e8-945b-a242f350a537.png)

**Comment by YesThatAllen** [comment]
Thanks for this @newhouse. Yes, I've confirmed what you said (more below). 

I was seeing the google URLs in the address bar, and the rest of the comments here added to the confusion, leading me to think that the google URL was still being visited. (Github allows comments to be dismissed as "resolved" which may help future visitors.

To confirm, I set my hosts file for `www.google.com` to resolve to `localhost`, loaded the official plugin, and tried to browse using the links from gmail.
They loaded fine.

I pasted a redirect URL in an incognito window, and they failed to load. So, yes, this plugin is stifling the `data-saferedirecturl` links. 

Very nice 👏

**Comment by aspiers** [comment]
Yeah thanks a lot @newhouse for your patience and very informative replies, and sorry for the late reply.

>You may see the URL flash in your browser bar or network history

This is what confused me, and based on the above clearly I wasn't the only one.  If there is a way to prevent that from appearing then it would prevent others being similarly confused in the future.  But I appreciate that from a technical standpoint that might be very difficult or even impossible to achieve.  Maybe instead it would be possible to briefly flash something up showing that the redirect was intercepted and skipped?

>but it will have been stopped from loading via a 307 Internal Redirect.

Thanks for the image showing this (especially considering I don't have Flash installed so can't view the videos).  How did you manage to get that request and redirect response appearing in the Network tab of Chrome's inspector?  When I clicked a link in my gmail which has this redirect, I only see a single HTTP POST to https://mail.google.com/sync/u/0/i/s?hl=en-GB&c=32 which gets a 200 response.

Just one other loose end to tie up:

[**@newhouse**](https://github.com/newhouse) commented on [28 Aug 2018, 15:23 BST](https://github.com/newhouse/url-tracking-stripper/issues/69#issuecomment-416604381 "2018-08-28T14:23:47Z - Replied by Github Reply Comments"):
>One thing I noticed in your example is that you use `link.to.other.site...` as your redirect example. That doesn't contain a protocol (e.g. 'https://') which I think will make that example not have the redirect skipped. In my experience, Gmail will include the protocol and if you _do_ include the protocol (`...q=https://link.to.other.site...`) then it works as expected.

I'm not sure but I suspect I accidentally removed the https:// protocol when I anonymising the target URL.  At least, all the gmail redirects I'm looking at now do include the protocol (`...q=https://link.to.other.site...`) like you say.

Thanks again!

**Comment by Alocyte** [comment]
<A style="TEXT-DECORATION: underline; COLOR: rgb(0,0,255)" href="https://techagilist.com/wp-content/OWA/?email=info@comms.alocyte.co.za" target=_blank data-saferedirecturl="https://www.google.com/url?q=http://seamus-fox.com/wordpress/wp-content/plugins/wf-wf/OWA/?email%3D%5B%5B-Email-%5D%5D&amp;source=gmail&amp;ust=1567455013529000&amp;usg=AFQjCNEFL_S_QRC-OSTlttzWXLrw_LOO7A">View Re-login and retrieve your pending messages.</A></B>
Its a real problem in emails used by spammers.

**Comment by gtabroker** [comment]
If a email mailing list sent me a link but I dont want them to know I clicked the link, is there anyway to find where it redirects to without them tracking my click?
